tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -953,66 +980,69 @@ void EnqueueProgramCommand::process() {
+    this->manager.cq_write(cmd.data(), wrap_packet_size_bytes, write_ptr);
+
+    this->manager.wrap_issue_queue_wr_ptr(this->command_queue_id);
 }
 
-// EnqueueWrapCommand section
-EnqueueWrapCommand::EnqueueWrapCommand(uint32_t command_queue_id, Device* device, SystemMemoryManager& manager, DeviceCommand::WrapRegion wrap_region) : command_queue_id(command_queue_id), manager(manager), wrap_region(wrap_region) {
-    this->device = device;
+EnqueueCompletionWrapCommand::EnqueueCompletionWrapCommand(uint32_t command_queue_id, Device* device, SystemMemoryManager& manager, uint32_t event) : EnqueueWrapCommand(command_queue_id, device, manager) {
+    this->event = event;
 }
 
-const DeviceCommand EnqueueWrapCommand::assemble_device_command(uint32_t) {
+const DeviceCommand EnqueueCompletionWrapCommand::assemble_device_command(uint32_t) {
     DeviceCommand command;
-    command.set_wrap(this->wrap_region);
+    command.set_wrap(DeviceCommand::WrapRegion::COMPLETION);
+    command.set_event(this->event);
     return command;
 }
 
-void EnqueueWrapCommand::process() {
+void EnqueueCompletionWrapCommand::process() {
     uint32_t write_ptr = this->manager.get_issue_queue_write_ptr(this->command_queue_id);
     uint32_t space_left_in_bytes = this->manager.get_issue_queue_limit(this->command_queue_id) - write_ptr;
-    // There may not be enough space in the issue queue to submit another command
-    // In that case we write as big of a vector as we can with the wrap index (0) set to wrap type
-    // To ensure that the issue queue write pointer does wrap, we need the wrap packet to be the full size of the issue queue
     uint32_t wrap_packet_size_bytes = std::min(space_left_in_bytes, DeviceCommand::NUM_BYTES_IN_DEVICE_COMMAND);
-
     const DeviceCommand cmd = this->assemble_device_command(0);
     this->manager.issue_queue_reserve_back(wrap_packet_size_bytes, this->command_queue_id);
     this->manager.cq_write(cmd.data(), wrap_packet_size_bytes, write_ptr);
-    if (this->wrap_region == DeviceCommand::WrapRegion::COMPLETION) {
-        // Wrap the read pointers for completion queue because device will start writing data at head of completion queue and there are no more reads to be done at current completion queue write pointer
-        // If we don't wrap the read then the subsequent read buffer command may attempt to read past the total command queue size
-        // because the read buffer command will see updated write pointer to compute num pages to read but the local read pointer is pointing to tail of completion queue
-        this->manager.wrap_completion_queue_rd_ptr(this->command_queue_id);
-        this->manager.issue_queue_push_back(wrap_packet_size_bytes, LAZY_COMMAND_QUEUE_MODE, this->command_queue_id);
-    } else {
-        this->manager.wrap_issue_queue_wr_ptr(this->command_queue_id);
-    }
+    this->manager.wrap_next_completion_queue_wr_ptr(this->command_queue_id);
```

Comment:
so host is keeping track of additional completion queue pointer? I don't know if I understand what "wrap next pointer" signifies

Reply:
hopefully was clearer after our code review meeting.